pkg/xcap/summary.go

```diff
@@ -0,0 +1,310 @@
+// leaving the original unchanged.
+type observations struct {
+	data map[StatisticKey]*AggregatedObservation
+}
+
+// newObservations creates an empty observations.
+func newObservations() *observations {
+	return &observations{data: make(map[StatisticKey]*AggregatedObservation)}
+}
+
+// filter returns a new observations containing only entries with matching stat keys.
+func (o *observations) filter(keys ...StatisticKey) *observations {
+	if len(keys) == 0 || o == nil {
+		return o
+	}
+
+	keySet := make(map[StatisticKey]struct{}, len(keys))
+	for _, k := range keys {
+		keySet[k] = struct{}{}
+	}
+
+	result := newObservations()
+	for k, obs := range o.data {
+		if _, ok := keySet[k]; ok {
+			result.data[k] = obs
+		}
+	}
+	return result
+}
+
+// prefix returns a new observations with all stat names prefixed.
+func (o *observations) prefix(p string) *observations {
+	if p == "" || o == nil {
+		return o
+	}
+
+	result := newObservations()
+	for k, obs := range o.data {
+		newKey := StatisticKey{
+			Name:        p + k.Name,
+			DataType:    k.DataType,
+			Aggregation: k.Aggregation,
+		}
+		result.data[newKey] = obs
+	}
+	return result
+}
+
+// normalizeKeys returns a new observations with dots replaced by underscores in stat names.
+func (o *observations) normalizeKeys() *observations {
+	if o == nil {
+		return o
+	}
+
+	result := newObservations()
+	for k, obs := range o.data {
+		newKey := StatisticKey{
+			Name:        strings.ReplaceAll(k.Name, ".", "_"),
+			DataType:    k.DataType,
+			Aggregation: k.Aggregation,
+		}
+		result.data[newKey] = obs
+	}
+	return result
+}
+
+// merge merges another observations into this one.
+func (o *observations) merge(other *observations) {
+	if other == nil {
+		return
+	}
+	for k, obs := range other.data {
+		if existing, ok := o.data[k]; ok {
+			existing.Merge(obs)
+		} else {
+			o.data[k] = &AggregatedObservation{
+				Statistic: obs.Statistic,
+				Value:     obs.Value,
+				Count:     obs.Count,
+			}
+		}
+	}
+}
+
+// ToLogValues converts observations to a slice suitable for go-kit/log.
+// Keys are sorted for deterministic output.
+func (o *observations) toLogValues() []any {
+	if o == nil {
+		return nil
+	}
+
+	// Collect key-value pairs for sorting by name.
+	type kv struct {
+		name  string
+		value any
+	}
+	pairs := make([]kv, 0, len(o.data))
+	for k, obs := range o.data {
+		pairs = append(pairs, kv{name: k.Name, value: obs.Value})
+	}
+	sort.Slice(pairs, func(i, j int) bool {
+		return strings.Compare(pairs[i].name, pairs[j].name) < 0
+	})
+
+	result := make([]any, 0, len(pairs)*2)
+	for _, p := range pairs {
+		value := p.value
+
+		// Format bytes values (keys ending with "_bytes")
+		if strings.HasSuffix(p.name, "_bytes") {
+			switch val := value.(type) {
+			case uint64:
+				value = humanize.Bytes(val)
+			case int64:
+				value = humanize.Bytes(uint64(val))
+			}
+		}
+
+		// Format duration values (keys ending with "duration_ns")
```

Comment:
Do these keys hold nanoseconds specifically? Prometheus-style metrics usually just use "seconds" as the unit so maybe we want to use that, rather than doing something different? 
I believe .String() on a duration already formats it into something like "3.23s" so the nanoseconds piece would get ignored anyway